File: frontend/apps/ios/Pantopus/Features/Shared/MapListHybrid/MapListHybridShell.swift
//
//  MapListHybridShell.swift
//  Pantopus
//
//  T6.6a (P24) — shared archetype for map+list hybrid surfaces.
//  Full-bleed `MapKit` layer underneath, five chrome slots overlaid
//  (top pill, category chips, map controls), and a draggable bottom
//  sheet that snaps between three detents (`.collapsed` / `.standard`
//  / `.expanded`). Future consumers: Gigs map (current `NearbyMapView`,
//  migrates in P26), Marketplace map mode, Discover Businesses map.
//
//  Decision context: docs/t6-open-questions-decisions.md Q9.
//

import CoreLocation
import MapKit
import SwiftUI

/// Shell for a full-bleed map paired with a 3-detent bottom sheet.
///
/// The shell owns:
/// - the MapKit canvas (rendering supplied `pins` + optional anchor disc)
/// - the sheet shell (rounded card, drag handle, drag-to-snap gesture)
/// - the chrome layout (top pill / category chips at top, map controls
///   anchored above the sheet edge)
///
/// The shell does **not** own:
/// - the pin model itself — consumers project their data into `MapPin`
/// - the camera — auto-fits on first load via the supplied anchor; if
///   richer camera control is needed, expose a `cameraPosition` binding
///   in a follow-up additive change
/// - any of the chrome content — every slot is a `@ViewBuilder` so
///   consumers supply their own back-pill, category strip, locate-me
///   stack, sheet header, and sheet body
///
/// Pin↔list sync is the consumer's job: when a pin is tapped the shell
/// calls `onPinTap(id)`; the consumer typically (a) updates its own
/// selection state and (b) snaps `detent` to `.standard` so the sheet
/// surfaces the matching card.
@MainActor
public struct MapListHybridShell<
    TopPill: View,
    CategoryChips: View,
    MapControlsContent: View,
    SheetHeader: View,
    SheetBody: View
>: View {
    private let pins: [MapPin]
    private let anchor: MapAnchor?
    private let selectedPinId: String?
    private let onPinTap: (String) -> Void
    @Binding private var detent: MapListHybridDetent

    private let topPill: () -> TopPill
    private let categoryChips: () -> CategoryChips
    private let mapControls: () -> MapControlsContent
    private let sheetHeader: () -> SheetHeader
    private let sheetBody: () -> SheetBody

    @State private var dragTranslation: CGFloat = 0
    @State private var cameraPosition: MapCameraPosition = .automatic
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    public init(
        pins: [MapPin],
        anchor: MapAnchor? = nil,
        selectedPinId: String? = nil,
        detent: Binding<MapListHybridDetent>,
        onPinTap: @escaping (String) -> Void = { _ in },
        @ViewBuilder topPill: @escaping () -> TopPill = { EmptyView() },
        @ViewBuilder categoryChips: @escaping () -> CategoryChips = { EmptyView() },
        @ViewBuilder mapControls: @escaping () -> MapControlsContent = { EmptyView() },
        @ViewBuilder sheetHeader: @escaping () -> SheetHeader = { EmptyView() },
        @ViewBuilder sheetBody: @escaping () -> SheetBody
    ) {
        self.pins = pins
        self.anchor = anchor
        self.selectedPinId = selectedPinId
        self._detent = detent
        self.onPinTap = onPinTap
        self.topPill = topPill
        self.categoryChips = categoryChips
        self.mapControls = mapControls
        self.sheetHeader = sheetHeader
        self.sheetBody = sheetBody
    }

    public var body: some View {
        GeometryReader { geo in
            let sheetHeight = max(120, detent.height + dragTranslation)
            ZStack(alignment: .top) {
                mapLayer
                topPill()
                    .padding(.top, geo.safeAreaInsets.top + 4)
                    .padding(.horizontal, 14)
                    .accessibilityIdentifier("mapListHybridTopPill")
                categoryChips()
                    .padding(.top, geo.safeAreaInsets.top + 50)
                    .accessibilityIdentifier("mapListHybridChips")
                mapControlsLayer(bottomInset: sheetHeight + 14)
                bottomSheet(height: sheetHeight)
            }
            .ignoresSafeArea(edges: .bottom)
        }
        .accessibilityIdentifier("mapListHybridShell")
        .onAppear {
            recenterCamera()
        }
        .onChange(of: anchor) { _, _ in
            recenterCamera()
        }
    }

    // MARK: - Map

    private var mapLayer: some View {
        Map(position: $cameraPosition, interactionModes: [.pan, .zoom]) {
            ForEach(pins) { pin in
                Annotation("", coordinate: pin.coordinate, anchor: .center) {
                    Button {
                        onPinTap(pin.id)
                    } label: {
                        MapListHybridPinDot(
                            pin: pin,
                            isActive: pin.id == selectedPinId,
                            reduceMotion: reduceMotion
                        )
                    }
                    .buttonStyle(.plain)
                    .accessibilityIdentifier("mapListHybridPin_\(pin.id)")
                    .accessibilityLabel("Map pin")
                }
            }
            if let anchor {
                Annotation("", coordinate: anchor.coordinate, anchor: .center) {
                    MapListHybridAnchorDot()
                        .accessibilityLabel("You are here")
                }
            }
        }
        .mapStyle(.standard(pointsOfInterest: .excludingAll))
        .accessibilityIdentifier("mapListHybridMap")
    }

    private func recenterCamera() {
        if let anchor {
            cameraPosition = .region(MKCoordinateRegion(
                center: anchor.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.024, longitudeDelta: 0.024)
            ))
        } else if let first = pins.first {
            cameraPosition = .region(MKCoordinateRegion(
                center: first.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.024, longitudeDelta: 0.024)
            ))
        }
    }

    // MARK: - Floating map controls

    private func mapControlsLayer(bottomInset: CGFloat) -> some View {
        mapControls()
            .padding(.trailing, 14)
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomTrailing)
            .padding(.bottom, bottomInset)
            .accessibilityIdentifier("mapListHybridMapControls")
    }

    // MARK: - Bottom sheet

    private func bottomSheet(height: CGFloat) -> some View {
        VStack(spacing: 0) {
            Spacer(minLength: 0)
            VStack(spacing: 0) {
                MapListHybridSheetGrabber()
                sheetHeader()
                    .accessibilityIdentifier("mapListHybridSheetHeader")
                sheetBody()
                    .frame(maxHeight: .infinity, alignment: .top)
                    .accessibilityIdentifier("mapListHybridSheetBody")
            }
            .frame(maxWidth: .infinity)
            .frame(height: height)
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
            .shadow(color: .black.opacity(0.12), radius: 10, x: 0, y: -10)
            .gesture(
                DragGesture()
                    .onChanged { gesture in
                        // `translation.height` is positive when the
                        // finger moves down. Negate so `dragTranslation`
                        // is positive when the sheet is being *grown*
                        // (finger moves up).
                        dragTranslation = -gesture.translation.height
                    }
                    .onEnded { gesture in
                        // Pass raw velocity (positive = downward) so the
                        // resolver semantics align with Android + web.
                        let velocity = gesture.predictedEndTranslation.height
                        let displacedHeight = detent.height + dragTranslation
                        let target = MapListHybridDetentResolver.resolve(
                            from: detent,
                            velocity: velocity,
                            displacedHeight: displacedHeight
                        )
                        withAnimation(snapAnimation) {
                            detent = target
                            dragTranslation = 0
                        }
                    }
            )
        }
        .accessibilityIdentifier("mapListHybridSheet")
        .accessibilityElement(children: .contain)
    }

    private var snapAnimation: Animation {
        if reduceMotion {
            return .linear(duration: 0.001)
        }
        return .interpolatingSpring(stiffness: 320, damping: 30)
    }
}

// MARK: - Sheet grabber

/// 40 × 4 pt rounded grabber tab. Lifted out so unit tests + previews
/// can match the design's pixel-exact handle without re-implementing it.
public struct MapListHybridSheetGrabber: View {
    public init() {}

    public var body: some View {
        Capsule()
            .fill(Theme.Color.appBorderStrong)
            .frame(width: 40, height: 4)
            .padding(.top, 8)
            .padding(.bottom, 4)
            .frame(maxWidth: .infinity)
            .contentShape(Rectangle())
            .accessibilityHidden(true)
    }
}

// MARK: - Pin glyph

/// Internal pin glyph. Mirrors the design — colored disc + white ring
/// for confirmed, dashed outline for pending, dual pulse halos for the
/// active selection (suppressed under reduce-motion).
struct MapListHybridPinDot: View {
    let pin: MapPin
    let isActive: Bool
    let reduceMotion: Bool

    @State private var pulse = false

    var body: some View {
        ZStack {
            if isActive && !reduceMotion {
                Circle()
                    .fill(pin.color.opacity(0.25))
                    .frame(width: 46, height: 46)
                    .scaleEffect(pulse ? 1.2 : 0.85)
                    .opacity(pulse ? 0 : 1)
                    .animation(
                        .easeOut(duration: 1.6).repeatForever(autoreverses: false),
                        value: pulse
                    )
                Circle()
                    .fill(pin.color.opacity(0.35))
                    .frame(width: 34, height: 34)
                    .scaleEffect(pulse ? 1.15 : 0.85)
                    .opacity(pulse ? 0 : 1)
                    .animation(
                        .easeOut(duration: 1.6).delay(0.4).repeatForever(autoreverses: false),
                        value: pulse
                    )
            } else if isActive {
                Circle()
                    .stroke(pin.color.opacity(0.4), lineWidth: 2)
                    .frame(width: 34, height: 34)
            }
            Circle()
                .fill(pin.color)
                .frame(width: 22, height: 22)
                .overlay(
                    Circle().stroke(
                        pin.state == .confirmed ? Color.white : Color.clear,
                        lineWidth: 2
                    )
                )
                .overlay(
                    Circle()
                        .strokeBorder(
                            pin.state == .pending ? pin.color : .clear,
                            style: StrokeStyle(lineWidth: 2, dash: [3, 2])
                        )
                        .scaleEffect(1.25)
                )
                .shadow(color: .black.opacity(0.30), radius: 2, x: 0, y: 2)
        }
        .frame(width: 50, height: 50)
        .onAppear { pulse = isActive && !reduceMotion }
        .onChange(of: isActive) { _, newValue in
            pulse = newValue && !reduceMotion
        }
    }
}

struct MapListHybridAnchorDot: View {
    var body: some View {
        ZStack {
            Circle()
                .fill(Theme.Color.primary600.opacity(0.18))
                .frame(width: 28, height: 28)
            Circle()
                .fill(Theme.Color.primary600)
                .frame(width: 14, height: 14)
                .overlay(Circle().stroke(Color.white, lineWidth: 3))
        }
    }
}

// MARK: - MapPin coordinate convenience

extension MapPin {
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}

extension MapAnchor {
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/map_list_hybrid/MapListHybridShell.kt
@file:Suppress("MagicNumber", "LongMethod", "PackageNaming", "CyclomaticComplexMethod")
@file:OptIn(com.google.maps.android.compose.MapsComposeExperimentalApi::class)

package app.pantopus.android.ui.screens.shared.map_list_hybrid

import android.content.Context
import android.provider.Settings
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.Orientation
import androidx.compose.foundation.gestures.draggable
import androidx.compose.foundation.gestures.rememberDraggableState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.Radii
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.maps.android.compose.CameraPositionState
import com.google.maps.android.compose.GoogleMap
import com.google.maps.android.compose.MapProperties
import com.google.maps.android.compose.MapUiSettings
import com.google.maps.android.compose.MarkerComposable
import com.google.maps.android.compose.MarkerState
import com.google.maps.android.compose.rememberCameraPositionState

/**
 * T6.6a (P24) — shared archetype for map+list hybrid surfaces.
 *
 * Full-bleed Google Maps canvas underneath, five chrome slots overlaid
 * (top pill, category chips at the top edge, map-controls stack pinned
 * to the right above the sheet edge), and a draggable bottom sheet
 * that snaps between three detents per Q9:
 * [MapListHybridDetent.Collapsed] (160 dp),
 * [MapListHybridDetent.Standard] (296 dp), and
 * [MapListHybridDetent.Expanded] (518 dp).
 *
 * The shell owns:
 * - the Google Maps canvas (rendering supplied [pins] + optional [anchor] disc)
 * - the sheet shell + drag-to-snap gesture
 * - the chrome layout
 *
 * The shell does **not** own data or chrome content — every slot is a
 * `@Composable` lambda so consumers supply their own back-pill, category
 * strip, locate-me / layers buttons, sheet header, and sheet body.
 *
 * Pin↔list sync is the consumer's job: when a pin is tapped the shell
 * fires [onPinTap]; the consumer typically updates its own selection
 * state and snaps the [detent] to `Standard` so the matching list item
 * surfaces.
 */
@Composable
fun MapListHybridShell(
    pins: List<MapPin>,
    detent: MapListHybridDetent,
    onDetentChange: (MapListHybridDetent) -> Unit,
    modifier: Modifier = Modifier,
    anchor: MapAnchor? = null,
    selectedPinId: String? = null,
    onPinTap: (String) -> Unit = {},
    reduceMotionOverride: Boolean? = null,
    topPill: @Composable () -> Unit = {},
    categoryChips: @Composable () -> Unit = {},
    mapControls: @Composable () -> Unit = {},
    sheetHeader: @Composable () -> Unit = {},
    sheetBody: @Composable () -> Unit = {},
) {
    val context = LocalContext.current
    val reduceMotion =
        reduceMotionOverride ?: remember(context) { systemReduceMotion(context) }
    val density = LocalDensity.current

    val cameraStartCoord =
        anchor?.let { LatLng(it.latitude, it.longitude) }
            ?: pins.firstOrNull()?.let { LatLng(it.latitude, it.longitude) }
            ?: LatLng(40.7484, -73.9857)

    val cameraState =
        rememberCameraPositionState {
            position = CameraPosition.fromLatLngZoom(cameraStartCoord, 15f)
        }

    var dragDelta by remember { mutableStateOf(0f) }
    val baseHeightDp = detent.height
    val baseHeightPx = with(density) { baseHeightDp.toPx() }
    val minHeightPx = 120f * density.density
    val targetHeightPx = (baseHeightPx - dragDelta).coerceAtLeast(minHeightPx)
    val targetHeightDp = with(density) { targetHeightPx.toDp() }
    val animatedHeight by animateDpAsState(
        targetValue = targetHeightDp,
        animationSpec = if (reduceMotion) tween(1) else tween(durationMillis = 240),
        label = "mapListHybridSheetHeight",
    )

    Box(
        modifier =
            modifier
                .fillMaxSize()
                .background(PantopusColors.appBg)
                .testTag("mapListHybridShell"),
    ) {
        MapLayer(
            cameraState = cameraState,
            pins = pins,
            selectedPinId = selectedPinId,
            anchor = anchor,
            reduceMotion = reduceMotion,
            onPinTap = onPinTap,
        )

        Box(
            modifier =
                Modifier
                    .padding(WindowInsets.statusBars.asPaddingValues())
                    .padding(top = 8.dp, start = 14.dp, end = 14.dp)
                    .align(Alignment.TopCenter)
                    .fillMaxWidth(),
        ) {
            topPill()
        }

        Box(
            modifier =
                Modifier
                    .padding(WindowInsets.statusBars.asPaddingValues())
                    .padding(top = 56.dp)
                    .align(Alignment.TopCenter)
                    .fillMaxWidth(),
        ) {
            categoryChips()
        }

        Box(
            modifier =
                Modifier
                    .align(Alignment.BottomEnd)
                    .padding(end = 14.dp, bottom = targetHeightDp + 14.dp)
                    .testTag("mapListHybridMapControls"),
        ) {
            mapControls()
        }

        BottomSheet(
            heightDp = animatedHeight,
            onDrag = { delta -> dragDelta += delta },
            onDragReleased = { velocity ->
                val displacedPx = baseHeightPx - dragDelta
                val targetsPx =
                    MapListHybridDetent.entries.associateWith { stop ->
                        with(density) { stop.height.toPx() }
                    }
                val next =
                    MapListHybridDetentResolver.resolve(
                        current = detent,
                        velocity = velocity,
                        displacedHeightPx = displacedPx,
                        targetsPx = targetsPx,
                    )
                onDetentChange(next)
                dragDelta = 0f
            },
            sheetHeader = sheetHeader,
            sheetBody = sheetBody,
            modifier = Modifier.align(Alignment.BottomCenter),
        )
    }
}

@Composable
private fun MapLayer(
    cameraState: CameraPositionState,
    pins: List<MapPin>,
    selectedPinId: String?,
    anchor: MapAnchor?,
    reduceMotion: Boolean,
    onPinTap: (String) -> Unit,
) {
    val mapProperties = remember { MapProperties(isMyLocationEnabled = false) }
    val uiSettings =
        remember {
            MapUiSettings(
                myLocationButtonEnabled = false,
                zoomControlsEnabled = false,
                compassEnabled = false,
                mapToolbarEnabled = false,
            )
        }
    GoogleMap(
        modifier = Modifier.fillMaxSize().testTag("mapListHybridMap"),
        cameraPositionState = cameraState,
        properties = mapProperties,
        uiSettings = uiSettings,
    ) {
        pins.forEach { pin ->
            val pinPosition = LatLng(pin.latitude, pin.longitude)
            val markerState =
                remember(pin.id, pin.latitude, pin.longitude) {
                    MarkerState(position = pinPosition)
                }
            val isActive = pin.id == selectedPinId
            MarkerComposable(
                keys = arrayOf<Any>(pin.id, isActive),
                state = markerState,
                anchor = androidx.compose.ui.geometry.Offset(0.5f, 0.5f),
                onClick = {
                    onPinTap(pin.id)
                    true
                },
            ) {
                MapListHybridPinDot(pin = pin, isActive = isActive, reduceMotion = reduceMotion)
            }
        }
        if (anchor != null) {
            val anchorState =
                remember(anchor.latitude, anchor.longitude) {
                    MarkerState(position = LatLng(anchor.latitude, anchor.longitude))
                }
            MarkerComposable(
                keys = arrayOf<Any>("anchor"),
                state = anchorState,
                anchor = androidx.compose.ui.geometry.Offset(0.5f, 0.5f),
            ) {
                MapListHybridAnchorDot()
            }
        }
    }
}

@Composable
private fun BottomSheet(
    heightDp: Dp,
    onDrag: (Float) -> Unit,
    onDragReleased: (Float) -> Unit,
    sheetHeader: @Composable () -> Unit,
    sheetBody: @Composable () -> Unit,
    modifier: Modifier = Modifier,
) {
    val draggable =
        rememberDraggableState { delta -> onDrag(delta) }
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .height(heightDp)
                .clip(RoundedCornerShape(topStart = 22.dp, topEnd = 22.dp))
                .background(PantopusColors.appSurface)
                .shadow(elevation = 10.dp, shape = RoundedCornerShape(topStart = 22.dp, topEnd = 22.dp))
                .draggable(
                    state = draggable,
                    orientation = Orientation.Vertical,
                    onDragStopped = { velocity -> onDragReleased(velocity) },
                )
                .testTag("mapListHybridSheet"),
    ) {
        MapListHybridSheetGrabber()
        sheetHeader()
        sheetBody()
    }
}

/** 40 × 4 dp pill grabber at the top of the sheet. */
@Composable
fun MapListHybridSheetGrabber(modifier: Modifier = Modifier) {
    Box(
        modifier =
            modifier
                .fillMaxWidth()
                .padding(top = 8.dp, bottom = 4.dp),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier =
                Modifier
                    .width(40.dp)
                    .height(4.dp)
                    .clip(RoundedCornerShape(Radii.xs))
                    .background(PantopusColors.appBorderStrong)
                    .semantics { contentDescription = "Drag handle" },
        )
    }
}

@Composable
internal fun MapListHybridPinDot(
    pin: MapPin,
    isActive: Boolean,
    reduceMotion: Boolean,
) {
    Box(
        modifier =
            Modifier
                .size(50.dp)
                .testTag("mapListHybridPin_${pin.id}"),
        contentAlignment = Alignment.Center,
    ) {
        // Static double-halo on active selection. Under reduce-motion
        // we swap the soft halo for a thin static ring so selection
        // remains visible without animating an infinite pulse.
        if (isActive) {
            Box(
                modifier =
                    Modifier
                        .size(46.dp)
                        .clip(CircleShape)
                        .background(pin.color.copy(alpha = if (reduceMotion) 0f else 0.25f))
                        .border(
                            width = if (reduceMotion) 2.dp else 0.dp,
                            color = if (reduceMotion) pin.color.copy(alpha = 0.45f) else Color.Transparent,
                            shape = CircleShape,
                        ),
            )
            Box(
                modifier =
                    Modifier
                        .size(34.dp)
                        .clip(CircleShape)
                        .background(pin.color.copy(alpha = if (reduceMotion) 0f else 0.35f)),
            )
        }
        Box(
            modifier =
                Modifier
                    .size(22.dp)
                    .clip(CircleShape)
                    .background(pin.color)
                    .border(
                        width = if (pin.state == MapPinState.Confirmed) 2.dp else 0.dp,
                        color = if (pin.state == MapPinState.Confirmed) Color.White else Color.Transparent,
                        shape = CircleShape,
                    )
                    .shadow(elevation = 2.dp, shape = CircleShape),
        )
        if (pin.state == MapPinState.Pending) {
            Box(
                modifier =
                    Modifier
                        .size(28.dp)
                        .border(2.dp, pin.color, CircleShape),
            )
        }
    }
}

@Composable
internal fun MapListHybridAnchorDot() {
    Box(
        modifier = Modifier.size(28.dp),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier =
                Modifier
                    .size(28.dp)
                    .clip(CircleShape)
                    .background(PantopusColors.primary600.copy(alpha = 0.18f)),
        )
        Box(
            modifier =
                Modifier
                    .size(14.dp)
                    .clip(CircleShape)
                    .background(PantopusColors.primary600)
                    .border(3.dp, Color.White, CircleShape)
                    .semantics { contentDescription = "You are here" },
        )
    }
}

/** Reads the device's transition-animation-scale to infer reduce-motion. */
internal fun systemReduceMotion(context: Context): Boolean =
    Settings.Global.getFloat(
        context.contentResolver,
        Settings.Global.TRANSITION_ANIMATION_SCALE,
        1f,
    ) == 0f
